src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,404 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+{
+    ucp_context_h UCS_V_UNUSED context = memh->context;
+    void *p                            = buffer;
+
+    /* Check that md_map is valid */
+    ucs_assertv(ucs_test_all_flags(UCS_MASK(context->num_mds), memh->md_map),
+                "mask 0x%lx memh %p md_map 0x%lx", UCS_MASK(context->num_mds),
+                memh, memh->md_map);
+
+    /* Size of mkey information which comes prior TL mkey data of all MDs */
+    ucp_memh_info_size_pack(&p, memh_info_size);
+
+    /* Flags */
+    *ucs_serialize_next(&p, uint16_t) = flags;
+
+    /* Memory domains map */
+    *ucs_serialize_next(&p, ucp_md_map_t) =
+            context->export_md_map[memh->mem_type];
+
+    /* Memory type */
+    UCS_STATIC_ASSERT(UCS_MEMORY_TYPE_LAST <= 255);
+    *ucs_serialize_next(&p, uint8_t) = memh->mem_type;
+
+    return p;
+}
+
+size_t ucp_memh_global_id_packed_size(uct_md_attr_v2_t *md_attr)
+{
+    size_t size = UCT_MD_GLOBAL_ID_MAX;
+
+    while ((size != 0) && (md_attr->global_id[size - 1] == '\0')) {
+        --size;
+    }
+
+    ucs_assert(size < UCT_MD_GLOBAL_ID_MAX);
+    return size;
+}
+
+static size_t
+ucp_memh_tl_mkey_common_packed_size(size_t specific_info_size)
+{
+    size_t size;
+
+    /* Size of packed TL mkey data */
+    size = sizeof(uint8_t);
+
+    /* TL mkey size */
+    size += sizeof(uint8_t);
+
+    return ucp_memh_extended_info_size(size, specific_info_size);
+}
+
+static size_t
+ucp_memh_exported_tl_mkey_packed_size(ucp_context_h context,
+                                      ucp_md_index_t md_index)
+{
+    uct_md_attr_v2_t *md_attr = &context->tl_mds[md_index].attr;
+    size_t size;
+
+    /* Size of exported TL mkey packed size */
+    size = md_attr->exported_mkey_packed_size;
+
+    /* Size of component name */
+    size = sizeof(uint8_t);
+
+    /* Component name */
+    size += strlen(md_attr->component_name);
+
+    /* Size of global MD identifier */
+    size += sizeof(uint8_t);
+
+    /* Global MD identifier */
+    size += ucp_memh_global_id_packed_size(md_attr);
+
+    return ucp_memh_tl_mkey_common_packed_size(size);
+}
+
+static size_t
+ucp_memh_exported_packed_size(ucp_context_h context, ucp_md_map_t md_map)
+{
+    size_t size = ucp_memh_exported_info_packed_size();
+    ucp_md_index_t md_index;
+
+    ucs_for_each_bit(md_index, md_map) {
+        size += ucp_memh_exported_tl_mkey_packed_size(context, md_index);
+    }
+
+    return size;
+}
+
+/**
+ *  memh_info_size_part1    :  8
+ * [memh_info_size_part2    : 16  - packed only if memh_info_size_part1 > 255]
+ *  flags                   : 16
+ *  md_map                  : 64
+ *  mem_type                :  8
+ *
+ * exported_memh :
+ * [
+ *  address          : 64
+ *  length           : 64
+ *  ucp_context_uuid : 64
+ *  registration_id  : 64
+ * ]
+ *
+ * ucs_for_each_bit(md_map) {
+ *     tl_mkey_data_size_part1 :  8
+ *    [tl_mkey_data_size_part2 : 16 - packed only if tl_mkey_data_size_part1 > 255]
```

Comment:
packed only if tl mkey data size > 255. In this case tl_mkey_data_size_part1 is set to 0